nexus/db-queries/src/db/datastore/ip_pool.rs

```diff
@@ -340,32 +378,60 @@ impl DataStore {
 
         // Find any pool of the given type linked to the caller's silo,
         // preferring default pools
-        ip_pool::table
+        let mut query = ip_pool::table
             .inner_join(ip_pool_resource::table)
             .filter(
                 ip_pool_resource::resource_type.eq(IpPoolResourceType::Silo),
             )
             .filter(ip_pool_resource::resource_id.eq(authz_silo_id))
             .filter(ip_pool::time_deleted.is_null())
             .filter(ip_pool::pool_type.eq(pool_type))
+            .into_boxed();
+
+        // Filter by IP version if specified
+        if let Some(version) = ip_version {
+            query = query.filter(ip_pool::ip_version.eq(version));
+        }
+
+        let pools: Vec<IpPool> = query
             .order((ip_pool_resource::is_default.desc(), ip_pool::name.asc()))
             .select(IpPool::as_select())
-            .first_async::<IpPool>(
-                &*self.pool_connection_authorized(opctx).await?,
-            )
+            .load_async(&*self.pool_connection_authorized(opctx).await?)
             .await
             .map_err(|e| {
                 public_error_from_diesel_lookup(
                     e,
                     ResourceType::IpPool,
                     &lookup_type,
                 )
-            })
-            .map(|ip_pool| {
+            })?;
+
+        // If `ip_version` was not specified and we have multiple pools of
+        // different IP versions, return an error asking the caller to specify
+        // which version.
+        if ip_version.is_none() && pools.len() > 1 {
```

Comment:
Might be worth factoring this out to a function, since it's used several places. Not critical though.